src/Language/Mimsa/Typechecker/Unify.hs

```diff
@@ -77,6 +78,18 @@ unify tyA tyB =
     (a, b) ->
       throwError $ UnificationError a b
 
+-- when combining records in branches of an If we need to check they are
+-- actually the same rather than combine them
+unifyStrict :: MonoType -> MonoType -> TcMonad Substitutions
```

Comment:
Starting to think this should always be the record behaviour, will come back to this though.